### PR TITLE
util/ctags.sh: a script for generating tags file with expanding macros

### PR DESCRIPTION
**Pull request by masatake**
This script requires Universal Ctags 6.0.0 or newer.

Usage:

	./util/ctags.sh [...options passed to ctags...]

Fixing #10251

### CODE REVIEW

**Comment by beldmit** [comment]
Is it also necessary to change the `make tags` definition?

**Comment by masatake** [comment]
> Is it also necessary to change the make tags definition?

Thank you. Updated.

File: util/ctags.sh [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    exit 1
+fi
+
+if ! "${CTAGS}" --version | grep -q "Universal Ctags"; then
+    echo "${CTAGS}: not Univrsal Ctags" 1>&2
+    exit 1
+fi
+
+if "${CTAGS}" --version | grep -q "Universal Ctags 5.*"; then
+    echo "${CTAGS}: Univrsal Ctags but too old (5)" 1>&2
```
Comment (mspncp) [inline util/ctags.sh:34]: Typo: Univrsal -> Universal 

```suggestion
    echo "${CTAGS}: Universal Ctags version is too old (5)" 1>&2
```

**Review by mspncp — COMMENTED** [review]
Just for my understanding: which feature of Universal Ctags exactly makes version 6.0.0 a requirement? 

Instead of just failing entirely, couldn't the script just gracefully fall back to a simpler behaviour if only an older version is available? In particular, since it is now linked to the `make tags` target?

File: util/ctags.sh [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+: ${CTAGS=ctags}
+: ${READTAGS=readtags}
+
+if ! type "${CTAGS}" > /dev/null; then
+    echo "${CTAGS}: not found" 1>&2
+    exit 1
+fi
+
+if ! "${CTAGS}" --version | grep -q "Universal Ctags"; then
+    echo "${CTAGS}: not Univrsal Ctags" 1>&2
```
[the remark this answers is not in the document]
Comment (masatake) [reply util/ctags.sh]: Added a commit for fixing the typo.

File: util/ctags.sh (hunk 1 above)
Comment (masatake) [reply util/ctags.sh]: Added a commit for fixing the typos.

**Comment by masatake** [comment]
Universal Ctags is derived from Exuberant Ctags 5.8.x.  Universal Ctags 6.0.0 is the first official release after forking.  Till releasing the version, Universal Ctags prints "5.9.0" when `--version` option is passed.  There is no reliable way to detect which version of Universal Ctags provides the "tagging after macro expanding" feature. We can say only "version 6.0.0 or later provides the feature".

**Comment by masatake** [comment]
Before merging, I would like the developers of OpenSSL to try the tags output emitted by the script.

Though the script solves #10251, it doesn't mean you are comfortable with the tags output.

The script runs ctags twice, once for extracting macro definitions. Using the readtags command, the script converts ALL the corrected macro definitions to the list of -D ctags options.
In the second pass, ctags expands the input before extracting tags.

ctags in the second pass doesn't consider #include and #undef when choosing a macro definition prepared with the -D options. That means the tag file may include entries extracted from unwanted macro expansions.

**Comment by paulidale** [comment]
I've been using the [earlier flavour of this script](https://github.com/universal-ctags/target-openssl/commit/23e8fb470f47ca00e27c0a7b03b2f8085bdaef5c) since it was mentioned.  It works a treat and has proven to be quite beneficial.

**Review by paulidale — APPROVED** [review]

**Review by mspncp — APPROVED** [review]
I'm currently using Exuberant Ctags 5.9 with Emacs. Didn't have the time yet to switch and test your script in practice, but I'm almost sure I'm going to love it. Thank you very much for your contribution!

@masatake would you mind to auto-squash your `fixup!` commit without moving the merge-base? Then @paulidale does not have to take care of it when  merging.

**Comment by paulidale** [comment]
Squashing on merge isn't a problem, we do it all the time.

File: .ctags.d/exclude.ctags [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+#
+# Licensed under the Apache License 2.0 (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+#
+
+# List file names or patterns you want ctags to ignore.
+--exclude=.ctags.d
+--exclude=test
```
[the remark this answers is not in the document]
Comment (masatake) [reply .ctags.d/exclude.ctags:12]: Thank you. I will add the directories for the submodules.

**Comment by masatake** [comment]
@mspncp

> I'm currently using Exuberant Ctags 5.9 with Emacs. 

Do you use TAGS files generated by etags that is a part of Emacs or TAGS files generated by ctags with -E option?

**Comment by mspncp** [comment]
> Do you use TAGS files generated by etags that is a part of Emacs or TAGS files generated by ctags with -E option?

I don't use the `etags` which is shipped with Emacs. Instead, I used Debian's `update-alternatives` mechanism to register Exuberant Ctags as `etags`, which to my understanding is equivalent to calling ctags with the  `-e` (lowercase) option, i.e. Emacs mode.
```
msp@debian:~$ ls -l /usr/bin/etags
lrwxrwxrwx 1 root root 23 25. Aug 2021  /usr/bin/etags -> /etc/alternatives/etags
msp@debian:~$ ls -l /etc/alternatives/etags
lrwxrwxrwx 1 root root 24  6. Sep 2021  /etc/alternatives/etags -> /usr/bin/ctags-exuberant
msp@debian:~$ ls -l /usr/bin/ctags-exuberant
-rwxr-xr-x 1 root root 296640 26. Dez 18:44 /usr/bin/ctags-exuberant
msp@debian:~$ etags --version
Exuberant Ctags 5.9~svn20110310, Copyright (C) 1996-2009 Darren Hiebert
  Addresses: <dhiebert@users.sourceforge.net>, http://ctags.sourceforge.net
  Optional compiled features: +wildcards, +regex
```

**Comment by t8m** [comment]
The force-push was just squashing.

File: .gitignore [hunk 4]
```diff
@@ -231,6 +231,7 @@ Makefile.save
 *.bak
 cscope.*
```
Comment (Gangstablue777) [inline .gitignore:232]: `
```suggestion
# -  cscope.***_[]()@masatake _**
```

**Pull request closed without merging**